src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -999,24 +1000,26 @@ impl LogicalJoin {
         }
     }
 
+    #[inline(never)]
     fn to_stream_temporal_join_with_index_selection(
         &self,
         predicate: EqJoinPredicate,
         ctx: &mut ToStreamContext,
-    ) -> Result<StreamTemporalJoin> {
+    ) -> Result<PlanRef> {
         // Index selection for temporal join.
         let right = self.right();
         // `should_be_temporal_join()` has already check right input for us.
         let logical_scan: &LogicalScan = right.as_logical_scan().unwrap();
 
         // Use primary table.
-        let mut result_plan = self.to_stream_temporal_join(predicate.clone(), ctx);
+        let mut result_plan: Result<StreamTemporalJoin> =
+            self.to_stream_temporal_join(predicate.clone(), ctx);
```

Comment:
I guess there's no difference to return a `PlanRef` or `StreamTemporalJoin`? 😄 IIUC, when appropriete optimization is on, the returned `StreamTemporalJoin` should be allocated in the frame of the caller.

Reply:
I though the same, but see https://github.com/risingwavelabs/risingwave/issues/21355#issuecomment-2808083547 , the stack size in release mode seems to be contrary to this assumption.

I also checked the assembly from `cargo asm`, after this change, the stack size indeed reduced a few hundred bytes...

Reply:
interesting, good to know that

Reply:
Does it mean that the plan node is somehow directly allocated in heap?

Reply:
Directly allocation in heap seems not possible:
- https://users.rust-lang.org/t/how-to-create-large-objects-directly-in-heap/26405
- https://users.rust-lang.org/t/is-there-equivalent-of-placement-new-in-rust/86611